subcommands/keys/rotate_targets.go

```diff
@@ -0,0 +1,124 @@
+package keys
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/foundriesio/fioctl/client"
+	"github.com/foundriesio/fioctl/subcommands"
+)
+
+func init() {
+	rotateTargets := &cobra.Command{
+		Use:   "rotate-targets <offline-creds.tgz>",
+		Short: "Rotate the offline target signing key for the Factory",
+		Run:   doRotateTargets,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			api = subcommands.Login(cmd)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	subcommands.RequireFactory(rotateTargets)
+	cmd.AddCommand(rotateTargets)
+}
+
+func doRotateTargets(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	credsFile := args[0]
+	assertWritable(credsFile)
+	creds := getOfflineCreds(credsFile)
+
+	root, err := api.TufRootGet(factory)
+	subcommands.DieNotNil(err)
+
+	// Target "rotation" works like this:
+	// 1. Find the "online target key" - this the key used by CI, so we don't
+	//    want to lose it.
+	// 2. Generate a new key.
+	// 3. Set these keys in root.json
+
+	onlineTargetId, err := findOnlineTargetId(factory)
+	subcommands.DieNotNil(err)
+
+	rootid, rootPk, err := findRoot(*root, creds)
+	fmt.Println("= Root keyid:", rootid)
+	subcommands.DieNotNil(err)
+	targetid, newCreds := addNewTargetKey(root, onlineTargetId, creds)
+	fmt.Println("= New target:", targetid)
+	subcommands.DieNotNil(signRoot(root, TufSigner{rootid, rootPk}))
+
+	tmpCreds := saveTempCreds(credsFile, newCreds)
+
+	bytes, err := json.MarshalIndent(root, "", "  ")
+	subcommands.DieNotNil(err)
+	fmt.Println("= Uploading new root")
+	body, err := api.TufRootPost(factory, bytes)
+	if err != nil {
+		fmt.Println("\nERROR:", err)
+		fmt.Println(body)
+		os.Exit(1)
+	}
+	if err := os.Rename(tmpCreds, credsFile); err != nil {
+		fmt.Println("\nERROR: Unable to update offline creds file.", err)
+		fmt.Println("Temp copy still available at:", tmpCreds)
+	}
+}
+
+func findOnlineTargetId(factory string) (string, error) {
+	meta, err := api.TargetsList(factory)
+	if err != nil {
+		return "", err
+	}
+	return meta.Signatures[0].KeyID, nil
```

Comment:
I'm not sure this is sane, at least this is an excessive call to an API.

Instead, I'd define an online key as the one, for which there is no key file in offline creds.
Well, to be more precise: during the first rotation, there will be just one targets key in root.json, and it is an online key.
During the subsequent rotations, there will be two keys, but only one of them will have a corresponding key file in creds - this is offline key, and the other one is online key.

Reply:
good point. this isn't fool proof. Unfortunately, the offline keys file includes the targets key used by CI, so we know it also. I've tried my best to reason about this all:

 https://github.com/foundriesio/fioctl/pull/81/commits/2918514620d37bd2e2a9dc3e400214a3a8e9c443